### 데이터 불러오기

In [1]:
import numpy as np
import pandas as pd

In [2]:
train = pd.read_csv('train.csv', index_col = 'no')
test = pd.read_csv('test.csv', index_col = 'no')

train

age          workclass  fnlwgt      education  education-num  \
no                                                                    
1       25            Private  219199           11th              7   
2       39            Private   52978   Some-college             10   
3       35            Private  196899      Bachelors             13   
4       64            Private  135527      Assoc-voc             11   
5       24            Private   60783   Some-college             10   
...    ...                ...     ...            ...            ...   
29301   20            Private  100605        HS-grad              9   
29302   21            Private  372636        HS-grad              9   
29303   18   Self-emp-not-inc  258474           10th              6   
29304   33            Private  157446           11th              7   
29305   65                  ?   94809        HS-grad              9   

            marital-status          occupation    relationship  \
no                                                               
1                 Divorced   Machine-op-inspct   Not-in-family   
2                 Divorced       Other-service   Not-in-family   
3            Never-married   Handlers-cleaners   Not-in-family   
4                 Divorced        Tech-support   Not-in-family   
5       Married-civ-spouse    Transport-moving         Husband   
...                    ...                 ...             ...   
29301        Never-married               Sales       Own-child   
29302        Never-married               Sales       Own-child   
29303        Never-married     Farming-fishing       Own-child   
29304        Never-married        Craft-repair   Not-in-family   
29305              Widowed                   ?   Not-in-family   

                      race      sex  capital-gain  capital-loss  \
no                                                                
1                    White     Male             0             0   
2                    White   Female             0          1721   
3       Asian-Pac-Islander   Female             0             0   
4                    White   Female             0             0   
5                    White     Male             0             0   
...                    ...      ...           ...           ...   
29301                Other     Male             0             0   
29302                Black     Male             0             0   
29303                White     Male             0             0   
29304                White     Male             0             0   
29305                White   Female             0             0   

       hours-per-week  native-country  income  
no                                             
1                  40   United-States       0  
2                  55   United-States       0  
3                  50           Haiti       0  
4                  40   United-States       0  
5                  70   United-States       1  
...               ...             ...     ...  
29301              40     Puerto-Rico       0  
29302              40   United-States       0  
29303              40   United-States       0  
29304              65   United-States       0  
29305              40   United-States       0  

[29305 rows x 15 columns]

In [3]:
test

age          workclass  fnlwgt      education  education-num  \
no                                                                    
29306   18                  ?  245274   Some-college             10   
29307   29            Private   83003        HS-grad              9   
29308   45            Private   35136      Bachelors             13   
29309   42   Self-emp-not-inc   64631      Bachelors             13   
29310   41            Private  195821      Doctorate             16   
...    ...                ...     ...            ...            ...   
48838   45   Self-emp-not-inc  116789        HS-grad              9   
48839   48            Private  185079        HS-grad              9   
48840   63            Private  117473   Some-college             10   
48841   18            Private  150817           11th              7   
48842   31            Private  341632     Assoc-acdm             12   

            marital-status        occupation    relationship    race      sex  \
no                                                                              
29306        Never-married                 ?       Own-child   White     Male   
29307   Married-civ-spouse     Other-service            Wife   White   Female   
29308   Married-civ-spouse      Tech-support         Husband   Black     Male   
29309   Married-civ-spouse   Exec-managerial         Husband   White     Male   
29310   Married-civ-spouse   Exec-managerial            Wife   White   Female   
...                    ...               ...             ...     ...      ...   
48838   Married-civ-spouse      Craft-repair         Husband   White     Male   
48839        Never-married   Exec-managerial   Not-in-family   White   Female   
48840   Married-civ-spouse    Prof-specialty         Husband   White     Male   
48841        Never-married             Sales       Own-child   White   Female   
48842   Married-civ-spouse      Tech-support         Husband   Black     Male   

       capital-gain  capital-loss  hours-per-week  native-country  
no                                                                 
29306             0             0              16   United-States  
29307             0             0              40   United-States  
29308             0             0              40   United-States  
29309             0             0              40   United-States  
29310             0          1902              40   United-States  
...             ...           ...             ...             ...  
48838             0             0              60   United-States  
48839             0             0              50   United-States  
48840          4386             0              40   United-States  
48841             0             0              20   United-States  
48842             0             0              46   United-States  

[19537 rows x 14 columns]

In [4]:
print(train.shape)
print(test.shape)

(29305, 15)
(19537, 14)


# 데이터 전처리

### 결측치 파악하기

범주형 결측치 파악

In [5]:
train['workclass'].value_counts()                 # workclass 결측치 있음, 4번째로 많은 범주에 해당

 Private             20410
 Self-emp-not-inc     2305
 Local-gov            1868
 ?                    1663
 State-gov            1201
 Self-emp-inc          987
 Federal-gov           854
 Without-pay            12
 Never-worked            5
Name: workclass, dtype: int64

In [6]:
test['workclass'].value_counts()         

 Private             13496
 Self-emp-not-inc     1557
 Local-gov            1268
 ?                    1136
 State-gov             780
 Self-emp-inc          708
 Federal-gov           578
 Without-pay             9
 Never-worked            5
Name: workclass, dtype: int64

In [7]:
train['education'].value_counts()

 HS-grad         9435
 Some-college    6558
 Bachelors       4848
 Masters         1587
 Assoc-voc       1230
 11th            1057
 Assoc-acdm       980
 10th             816
 7th-8th          574
 Prof-school      500
 9th              430
 12th             415
 Doctorate        347
 5th-6th          334
 1st-4th          148
 Preschool         46
Name: education, dtype: int64

In [8]:
train['marital-status'].value_counts()

 Married-civ-spouse       13331
 Never-married             9673
 Divorced                  4030
 Widowed                    942
 Separated                  914
 Married-spouse-absent      394
 Married-AF-spouse           21
Name: marital-status, dtype: int64

In [9]:
train['occupation'].value_counts()                  # occupation  결측치 있음, 중간정도로 많은 범주에 해당

 Prof-specialty       3724
 Craft-repair         3632
 Exec-managerial      3609
 Adm-clerical         3440
 Sales                3292
 Other-service        2975
 Machine-op-inspct    1811
 ?                    1668
 Transport-moving     1393
 Handlers-cleaners    1249
 Farming-fishing       888
 Tech-support          865
 Protective-serv       603
 Priv-house-serv       146
 Armed-Forces           10
Name: occupation, dtype: int64

In [10]:
test['occupation'].value_counts()                  # occupation  결측치 있음, 중간정도로 많은 범주에 해당

 Craft-repair         2480
 Exec-managerial      2477
 Prof-specialty       2448
 Sales                2212
 Adm-clerical         2171
 Other-service        1948
 Machine-op-inspct    1211
 ?                    1141
 Transport-moving      962
 Handlers-cleaners     823
 Farming-fishing       602
 Tech-support          581
 Protective-serv       380
 Priv-house-serv        96
 Armed-Forces            5
Name: occupation, dtype: int64

In [11]:
train['relationship'].value_counts()

 Husband           11769
 Not-in-family      7637
 Own-child          4554
 Unmarried          3071
 Wife               1370
 Other-relative      904
Name: relationship, dtype: int64

In [12]:
train['race'].value_counts()                      

 White                 25081
 Black                  2776
 Asian-Pac-Islander      919
 Amer-Indian-Eskimo      290
 Other                   239
Name: race, dtype: int64

In [13]:
train['sex'].value_counts()

 Male      19621
 Female     9684
Name: sex, dtype: int64

In [14]:
train['native-country'].value_counts()             # native-country 결측치 있음, 3번째로 많은 범주에 해당

 United-States                 26301
 Mexico                          579
 ?                               495
 Philippines                     178
 Germany                         124
 Puerto-Rico                     117
 Canada                          111
 El-Salvador                      96
 Cuba                             94
 India                            90
 England                          75
 South                            74
 China                            68
 Jamaica                          68
 Dominican-Republic               61
 Italy                            60
 Japan                            54
 Poland                           53
 Columbia                         53
 Haiti                            51
 Guatemala                        51
 Vietnam                          50
 Portugal                         43
 Taiwan                           39
 Iran                             38
 Nicaragua                        28
 Greece                           25
 

In [15]:
test['native-country'].value_counts()             # native-country 결측치 있음, 3번째로 많은 범주에 해당

 United-States                 17531
 Mexico                          372
 ?                               362
 Philippines                     117
 Germany                          82
 Canada                           71
 Puerto-Rico                      67
 India                            61
 El-Salvador                      59
 China                            54
 England                          52
 Italy                            45
 Cuba                             44
 Dominican-Republic               42
 South                            41
 Japan                            38
 Jamaica                          38
 Guatemala                        37
 Vietnam                          36
 Poland                           34
 Columbia                         32
 Taiwan                           26
 Portugal                         24
 Ecuador                          24
 Greece                           24
 Haiti                            24
 Iran                             21
 

In [16]:
train['income'].value_counts()

0    22260
1     7045
Name: income, dtype: int64

- 숫자형 데이터 결측치 파악

In [17]:
train

age          workclass  fnlwgt      education  education-num  \
no                                                                    
1       25            Private  219199           11th              7   
2       39            Private   52978   Some-college             10   
3       35            Private  196899      Bachelors             13   
4       64            Private  135527      Assoc-voc             11   
5       24            Private   60783   Some-college             10   
...    ...                ...     ...            ...            ...   
29301   20            Private  100605        HS-grad              9   
29302   21            Private  372636        HS-grad              9   
29303   18   Self-emp-not-inc  258474           10th              6   
29304   33            Private  157446           11th              7   
29305   65                  ?   94809        HS-grad              9   

            marital-status          occupation    relationship  \
no                                                               
1                 Divorced   Machine-op-inspct   Not-in-family   
2                 Divorced       Other-service   Not-in-family   
3            Never-married   Handlers-cleaners   Not-in-family   
4                 Divorced        Tech-support   Not-in-family   
5       Married-civ-spouse    Transport-moving         Husband   
...                    ...                 ...             ...   
29301        Never-married               Sales       Own-child   
29302        Never-married               Sales       Own-child   
29303        Never-married     Farming-fishing       Own-child   
29304        Never-married        Craft-repair   Not-in-family   
29305              Widowed                   ?   Not-in-family   

                      race      sex  capital-gain  capital-loss  \
no                                                                
1                    White     Male             0             0   
2                    White   Female             0          1721   
3       Asian-Pac-Islander   Female             0             0   
4                    White   Female             0             0   
5                    White     Male             0             0   
...                    ...      ...           ...           ...   
29301                Other     Male             0             0   
29302                Black     Male             0             0   
29303                White     Male             0             0   
29304                White     Male             0             0   
29305                White   Female             0             0   

       hours-per-week  native-country  income  
no                                             
1                  40   United-States       0  
2                  55   United-States       0  
3                  50           Haiti       0  
4                  40   United-States       0  
5                  70   United-States       1  
...               ...             ...     ...  
29301              40     Puerto-Rico       0  
29302              40   United-States       0  
29303              40   United-States       0  
29304              65   United-States       0  
29305              40   United-States       0  

[29305 rows x 15 columns]

In [18]:
is_question = train['age'] == " ?"
train[is_question]                           # 모든 숫자형 컬럼 확인 결과 결측치 없음!!

C:\Users\rnjsx\anaconda3\lib\site-packages\pandas\core\ops\array_ops.py:253: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  res_values = method(rvalues)


Empty DataFrame
Columns: [age, workclass, fnlwgt, education, education-num, marital-status, occupation, relationship, race, sex, capital-gain, capital-loss, hours-per-week, native-country, income]
Index: []

workclass : 1663, 
occupation : 1668, 
native-country : 495 결측치 존재

### 결측치 채우기 2번 방법 : 그대로 사용하기 (Others로 분류)

- workclass 결측치 채우기

In [19]:
train['workclass'] = train['workclass'].replace(" ?", "Others")

In [20]:
train['workclass'].value_counts()                     # train의 workclass 결측치: ?  -> Others로 변경 됐다!

 Private             20410
 Self-emp-not-inc     2305
 Local-gov            1868
Others                1663
 State-gov            1201
 Self-emp-inc          987
 Federal-gov           854
 Without-pay            12
 Never-worked            5
Name: workclass, dtype: int64

In [21]:
test['workclass'] = test['workclass'].replace(" ?", "Others")

In [22]:
test['workclass'].value_counts()                     # test의 workclass 결측치: ?  -> Others로 변경 됐다!

 Private             13496
 Self-emp-not-inc     1557
 Local-gov            1268
Others                1136
 State-gov             780
 Self-emp-inc          708
 Federal-gov           578
 Without-pay             9
 Never-worked            5
Name: workclass, dtype: int64

- occupation 결측치 채우기

In [23]:
train['occupation'] = train['occupation'].replace(" ?", "Others")

In [24]:
train['occupation'].value_counts()                         # occupation 결측치: ?  -> Others로 변경 됐다!

 Prof-specialty       3724
 Craft-repair         3632
 Exec-managerial      3609
 Adm-clerical         3440
 Sales                3292
 Other-service        2975
 Machine-op-inspct    1811
Others                1668
 Transport-moving     1393
 Handlers-cleaners    1249
 Farming-fishing       888
 Tech-support          865
 Protective-serv       603
 Priv-house-serv       146
 Armed-Forces           10
Name: occupation, dtype: int64

In [25]:
test['occupation'] = test['occupation'].replace(" ?", "Others")

In [26]:
test['occupation'].value_counts()                         # test의 occupation 결측치: ?  -> Others로 변경 됐다!

 Craft-repair         2480
 Exec-managerial      2477
 Prof-specialty       2448
 Sales                2212
 Adm-clerical         2171
 Other-service        1948
 Machine-op-inspct    1211
Others                1141
 Transport-moving      962
 Handlers-cleaners     823
 Farming-fishing       602
 Tech-support          581
 Protective-serv       380
 Priv-house-serv        96
 Armed-Forces            5
Name: occupation, dtype: int64

- native-country 결측치 채우기

In [27]:
train['native-country'] = train['native-country'].replace(" ?", "Others")

In [28]:
train['native-country'].value_counts()                           # native-country 결측치: ?  -> Others로 변경 됐다!

 United-States                 26301
 Mexico                          579
Others                           495
 Philippines                     178
 Germany                         124
 Puerto-Rico                     117
 Canada                          111
 El-Salvador                      96
 Cuba                             94
 India                            90
 England                          75
 South                            74
 Jamaica                          68
 China                            68
 Dominican-Republic               61
 Italy                            60
 Japan                            54
 Poland                           53
 Columbia                         53
 Haiti                            51
 Guatemala                        51
 Vietnam                          50
 Portugal                         43
 Taiwan                           39
 Iran                             38
 Nicaragua                        28
 Greece                           25
 

In [29]:
test['native-country'] = test['native-country'].replace(" ?", "Others")

In [30]:
test['native-country'].value_counts()                           # test의 native-country 결측치: ?  -> Others로 변경 됐다!

 United-States                 17531
 Mexico                          372
Others                           362
 Philippines                     117
 Germany                          82
 Canada                           71
 Puerto-Rico                      67
 India                            61
 El-Salvador                      59
 China                            54
 England                          52
 Italy                            45
 Cuba                             44
 Dominican-Republic               42
 South                            41
 Japan                            38
 Jamaica                          38
 Guatemala                        37
 Vietnam                          36
 Poland                           34
 Columbia                         32
 Taiwan                           26
 Portugal                         24
 Ecuador                          24
 Greece                           24
 Haiti                            24
 Iran                             21
 

In [31]:
train

age          workclass  fnlwgt      education  education-num  \
no                                                                    
1       25            Private  219199           11th              7   
2       39            Private   52978   Some-college             10   
3       35            Private  196899      Bachelors             13   
4       64            Private  135527      Assoc-voc             11   
5       24            Private   60783   Some-college             10   
...    ...                ...     ...            ...            ...   
29301   20            Private  100605        HS-grad              9   
29302   21            Private  372636        HS-grad              9   
29303   18   Self-emp-not-inc  258474           10th              6   
29304   33            Private  157446           11th              7   
29305   65             Others   94809        HS-grad              9   

            marital-status          occupation    relationship  \
no                                                               
1                 Divorced   Machine-op-inspct   Not-in-family   
2                 Divorced       Other-service   Not-in-family   
3            Never-married   Handlers-cleaners   Not-in-family   
4                 Divorced        Tech-support   Not-in-family   
5       Married-civ-spouse    Transport-moving         Husband   
...                    ...                 ...             ...   
29301        Never-married               Sales       Own-child   
29302        Never-married               Sales       Own-child   
29303        Never-married     Farming-fishing       Own-child   
29304        Never-married        Craft-repair   Not-in-family   
29305              Widowed              Others   Not-in-family   

                      race      sex  capital-gain  capital-loss  \
no                                                                
1                    White     Male             0             0   
2                    White   Female             0          1721   
3       Asian-Pac-Islander   Female             0             0   
4                    White   Female             0             0   
5                    White     Male             0             0   
...                    ...      ...           ...           ...   
29301                Other     Male             0             0   
29302                Black     Male             0             0   
29303                White     Male             0             0   
29304                White     Male             0             0   
29305                White   Female             0             0   

       hours-per-week  native-country  income  
no                                             
1                  40   United-States       0  
2                  55   United-States       0  
3                  50           Haiti       0  
4                  40   United-States       0  
5                  70   United-States       1  
...               ...             ...     ...  
29301              40     Puerto-Rico       0  
29302              40   United-States       0  
29303              40   United-States       0  
29304              65   United-States       0  
29305              40   United-States       0  

[29305 rows x 15 columns]

train, test 데이터 모두 교체 완료!

In [32]:
train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 29305 entries, 1 to 29305
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             29305 non-null  int64 
 1   workclass       29305 non-null  object
 2   fnlwgt          29305 non-null  int64 
 3   education       29305 non-null  object
 4   education-num   29305 non-null  int64 
 5   marital-status  29305 non-null  object
 6   occupation      29305 non-null  object
 7   relationship    29305 non-null  object
 8   race            29305 non-null  object
 9   sex             29305 non-null  object
 10  capital-gain    29305 non-null  int64 
 11  capital-loss    29305 non-null  int64 
 12  hours-per-week  29305 non-null  int64 
 13  native-country  29305 non-null  object
 14  income          29305 non-null  int64 
dtypes: int64(7), object(8)
memory usage: 3.6+ MB


In [33]:
test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19537 entries, 29306 to 48842
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             19537 non-null  int64 
 1   workclass       19537 non-null  object
 2   fnlwgt          19537 non-null  int64 
 3   education       19537 non-null  object
 4   education-num   19537 non-null  int64 
 5   marital-status  19537 non-null  object
 6   occupation      19537 non-null  object
 7   relationship    19537 non-null  object
 8   race            19537 non-null  object
 9   sex             19537 non-null  object
 10  capital-gain    19537 non-null  int64 
 11  capital-loss    19537 non-null  int64 
 12  hours-per-week  19537 non-null  int64 
 13  native-country  19537 non-null  object
dtypes: int64(6), object(8)
memory usage: 2.2+ MB


In [34]:
train['workclass'].unique()

array([' Private', ' Self-emp-not-inc', ' Local-gov', 'Others',
       ' State-gov', ' Self-emp-inc', ' Federal-gov', ' Without-pay',
       ' Never-worked'], dtype=object)

In [35]:
test['workclass'].unique()

array(['Others', ' Private', ' Self-emp-not-inc', ' Local-gov',
       ' Self-emp-inc', ' State-gov', ' Federal-gov', ' Without-pay',
       ' Never-worked'], dtype=object)

# capital gain loss 전처리 2 (gain 0 / 소 / 대) - 평균기준

## gain 평균 및 범위 구하기

In [36]:
train[train['capital-gain'] == 1]

Empty DataFrame
Columns: [age, workclass, fnlwgt, education, education-num, marital-status, occupation, relationship, race, sex, capital-gain, capital-loss, hours-per-week, native-country, income]
Index: []

1인 데이터 없으니 range로 써도 되겠다

In [37]:
train[train['capital-gain'] > 0].describe()

age        fnlwgt  education-num  capital-gain  capital-loss  \
count  2422.000000  2.422000e+03    2422.000000   2422.000000        2422.0   
mean     44.006606  1.887691e+05      11.036334  13122.297275           0.0   
std      13.370177  1.043361e+05       2.681711  22848.391387           0.0   
min      17.000000  1.930200e+04       1.000000    114.000000           0.0   
25%      35.000000  1.173105e+05       9.000000   3411.000000           0.0   
50%      43.000000  1.760985e+05      10.000000   7073.500000           0.0   
75%      52.000000  2.394725e+05      13.000000  13550.000000           0.0   
max      90.000000  1.033222e+06      16.000000  99999.000000           0.0   

       hours-per-week       income  
count     2422.000000  2422.000000  
mean        43.446738     0.611065  
std         12.573791     0.487609  
min          1.000000     0.000000  
25%         40.000000     0.000000  
50%         40.000000     1.000000  
75%         50.000000     1.000000  
max         99.000000     1.000000

0이 아닌 capital-gain 값들의 평균은 13122 / 중간값은 7073이네

In [38]:
test[test['capital-gain'] == 1]

Empty DataFrame
Columns: [age, workclass, fnlwgt, education, education-num, marital-status, occupation, relationship, race, sex, capital-gain, capital-loss, hours-per-week, native-country]
Index: []

1인 데이터 없으니 range로 써도 되겠다

## gain 합치기 - 중앙

In [39]:
bins = [0, 1, 7073, 99999] # 범주값
labels = ['Zero', 'Small', 'Large'] # Alone 0-1 / Small 1-4 / Large 4-11

train['capital-gain-group'] = pd.cut(train['capital-gain'], bins=bins, labels=labels)
test['capital-gain-group'] = pd.cut(test['capital-gain'], bins=bins, labels=labels)

In [40]:
train['capital-gain']

no
1        0
2        0
3        0
4        0
5        0
        ..
29301    0
29302    0
29303    0
29304    0
29305    0
Name: capital-gain, Length: 29305, dtype: int64

In [41]:
train['capital-gain-group']

no
1        NaN
2        NaN
3        NaN
4        NaN
5        NaN
        ... 
29301    NaN
29302    NaN
29303    NaN
29304    NaN
29305    NaN
Name: capital-gain-group, Length: 29305, dtype: category
Categories (3, object): [Zero < Small < Large]

In [42]:
train['capital-gain-group'].unique()

[NaN, Large, Small]
Categories (2, object): [Small < Large]

In [43]:
train['capital-gain-group'] = train['capital-gain-group'].fillna('Zero')

In [44]:
train['capital-gain-group'].unique()

[Zero, Large, Small]
Categories (3, object): [Zero < Small < Large]

## loss 평균 및 범위 구하기

In [45]:
train[train['capital-loss'] == 1]

Empty DataFrame
Columns: [age, workclass, fnlwgt, education, education-num, marital-status, occupation, relationship, race, sex, capital-gain, capital-loss, hours-per-week, native-country, income, capital-gain-group]
Index: []

1인 데이터 없으니 range로 써도 되겠다

In [46]:
train[train['capital-loss'] > 0].describe()

age         fnlwgt  education-num  capital-gain  capital-loss  \
count  1369.000000    1369.000000    1369.000000        1369.0   1369.000000   
mean     41.641344  189810.688093      11.010226           0.0   1863.908692   
std      13.005535  101268.612368       2.601430           0.0    366.229158   
min      17.000000   20953.000000       2.000000           0.0    213.000000   
25%      32.000000  119904.000000       9.000000           0.0   1672.000000   
50%      41.000000  178319.000000      10.000000           0.0   1887.000000   
75%      50.000000  236543.000000      13.000000           0.0   1977.000000   
max      90.000000  816750.000000      16.000000           0.0   4356.000000   

       hours-per-week       income  
count     1369.000000  1369.000000  
mean        43.344777     0.476991  
std         12.170489     0.499653  
min          1.000000     0.000000  
25%         40.000000     0.000000  
50%         40.000000     0.000000  
75%         50.000000     1.000000  
max         99.000000     1.000000

0이 아닌 capital-loss 값들의 평균은 1863 / 중간값은 1887이네

In [47]:
test[test['capital-loss'] == 1]

Empty DataFrame
Columns: [age, workclass, fnlwgt, education, education-num, marital-status, occupation, relationship, race, sex, capital-gain, capital-loss, hours-per-week, native-country, capital-gain-group]
Index: []

1인 데이터 없으니 range로 써도 되겠다

## loss 합치기 - 평균

In [48]:
bins = [0, 1, 1887, 99999] # 범주값
labels = ['Zero', 'Small', 'Large'] # Alone 0-1 / Small 1-4 / Large 4-11

train['capital-loss-group'] = pd.cut(train['capital-loss'], bins=bins, labels=labels)
test['capital-loss-group'] = pd.cut(test['capital-loss'], bins=bins, labels=labels)

In [49]:
train['capital-loss-group']

no
1          NaN
2        Small
3          NaN
4          NaN
5          NaN
         ...  
29301      NaN
29302      NaN
29303      NaN
29304      NaN
29305      NaN
Name: capital-loss-group, Length: 29305, dtype: category
Categories (3, object): [Zero < Small < Large]

In [50]:
train['capital-loss-group'] = train['capital-loss-group'].fillna('Zero')

In [51]:
train['capital-loss-group'].unique()

[Zero, Small, Large]
Categories (3, object): [Zero < Small < Large]

## 원본 capital gain, loss 삭제

In [52]:
train.drop('capital-gain', axis = 1, inplace=True)

In [53]:
train.drop('capital-loss', axis = 1, inplace=True)

In [54]:
train.columns

Index(['age', 'workclass', 'fnlwgt', 'education', 'education-num',
       'marital-status', 'occupation', 'relationship', 'race', 'sex',
       'hours-per-week', 'native-country', 'income', 'capital-gain-group',
       'capital-loss-group'],
      dtype='object')

In [55]:
test.drop('capital-gain', axis = 1, inplace=True)

In [56]:
test.drop('capital-loss', axis = 1, inplace=True)

In [57]:
test.columns

Index(['age', 'workclass', 'fnlwgt', 'education', 'education-num',
       'marital-status', 'occupation', 'relationship', 'race', 'sex',
       'hours-per-week', 'native-country', 'capital-gain-group',
       'capital-loss-group'],
      dtype='object')

### 데이터 컬럼 병합 및 삭제

In [58]:
test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19537 entries, 29306 to 48842
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype   
---  ------              --------------  -----   
 0   age                 19537 non-null  int64   
 1   workclass           19537 non-null  object  
 2   fnlwgt              19537 non-null  int64   
 3   education           19537 non-null  object  
 4   education-num       19537 non-null  int64   
 5   marital-status      19537 non-null  object  
 6   occupation          19537 non-null  object  
 7   relationship        19537 non-null  object  
 8   race                19537 non-null  object  
 9   sex                 19537 non-null  object  
 10  hours-per-week      19537 non-null  int64   
 11  native-country      19537 non-null  object  
 12  capital-gain-group  1613 non-null   category
 13  capital-loss-group  913 non-null    category
dtypes: category(2), int64(4), object(8)
memory usage: 2.0+ MB


# 모델 예측

### 문제, 정답 데이터로 분리

In [59]:
y_train = train['income']
X_train = train.drop('income', axis=1)
X_test = test

In [60]:
X_train.columns

Index(['age', 'workclass', 'fnlwgt', 'education', 'education-num',
       'marital-status', 'occupation', 'relationship', 'race', 'sex',
       'hours-per-week', 'native-country', 'capital-gain-group',
       'capital-loss-group'],
      dtype='object')

In [61]:
X_test.columns

Index(['age', 'workclass', 'fnlwgt', 'education', 'education-num',
       'marital-status', 'occupation', 'relationship', 'race', 'sex',
       'hours-per-week', 'native-country', 'capital-gain-group',
       'capital-loss-group'],
      dtype='object')

In [62]:
train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 29305 entries, 1 to 29305
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype   
---  ------              --------------  -----   
 0   age                 29305 non-null  int64   
 1   workclass           29305 non-null  object  
 2   fnlwgt              29305 non-null  int64   
 3   education           29305 non-null  object  
 4   education-num       29305 non-null  int64   
 5   marital-status      29305 non-null  object  
 6   occupation          29305 non-null  object  
 7   relationship        29305 non-null  object  
 8   race                29305 non-null  object  
 9   sex                 29305 non-null  object  
 10  hours-per-week      29305 non-null  int64   
 11  native-country      29305 non-null  object  
 12  income              29305 non-null  int64   
 13  capital-gain-group  29305 non-null  category
 14  capital-loss-group  29305 non-null  category
dtypes: category(2), int64(5), object(8)


### 데이터 원핫 인코딩 하기

In [63]:
X_train = pd.get_dummies(X_train)
X_test = pd.get_dummies(X_test)

In [64]:
print(X_train.shape)
print(X_test.shape)

(29305, 112)
(19537, 111)


### 컬럼 개수 맞추기

In [65]:
set(X_train.columns) - set(X_test.columns)

{'native-country_ Holand-Netherlands'}

In [66]:
# 'native-country_ Holand-Netherlands'의 컬럼은 생기지만 실제로 컴퓨터가 데이터는 없는 것으로 인식함!
X_test['native-country_ Holand-Netherlands'] = 0

In [67]:
tmp = X_train.columns
for x in tmp :
    print(x)

age
fnlwgt
education-num
hours-per-week
workclass_ Federal-gov
workclass_ Local-gov
workclass_ Never-worked
workclass_ Private
workclass_ Self-emp-inc
workclass_ Self-emp-not-inc
workclass_ State-gov
workclass_ Without-pay
workclass_Others
education_ 10th
education_ 11th
education_ 12th
education_ 1st-4th
education_ 5th-6th
education_ 7th-8th
education_ 9th
education_ Assoc-acdm
education_ Assoc-voc
education_ Bachelors
education_ Doctorate
education_ HS-grad
education_ Masters
education_ Preschool
education_ Prof-school
education_ Some-college
marital-status_ Divorced
marital-status_ Married-AF-spouse
marital-status_ Married-civ-spouse
marital-status_ Married-spouse-absent
marital-status_ Never-married
marital-status_ Separated
marital-status_ Widowed
occupation_ Adm-clerical
occupation_ Armed-Forces
occupation_ Craft-repair
occupation_ Exec-managerial
occupation_ Farming-fishing
occupation_ Handlers-cleaners
occupation_ Machine-op-inspct
occupation_ Other-service
occupation_ Priv-hou

In [68]:
tmp = X_test.columns
for x in tmp :
    print(x)

age
fnlwgt
education-num
hours-per-week
workclass_ Federal-gov
workclass_ Local-gov
workclass_ Never-worked
workclass_ Private
workclass_ Self-emp-inc
workclass_ Self-emp-not-inc
workclass_ State-gov
workclass_ Without-pay
workclass_Others
education_ 10th
education_ 11th
education_ 12th
education_ 1st-4th
education_ 5th-6th
education_ 7th-8th
education_ 9th
education_ Assoc-acdm
education_ Assoc-voc
education_ Bachelors
education_ Doctorate
education_ HS-grad
education_ Masters
education_ Preschool
education_ Prof-school
education_ Some-college
marital-status_ Divorced
marital-status_ Married-AF-spouse
marital-status_ Married-civ-spouse
marital-status_ Married-spouse-absent
marital-status_ Never-married
marital-status_ Separated
marital-status_ Widowed
occupation_ Adm-clerical
occupation_ Armed-Forces
occupation_ Craft-repair
occupation_ Exec-managerial
occupation_ Farming-fishing
occupation_ Handlers-cleaners
occupation_ Machine-op-inspct
occupation_ Other-service
occupation_ Priv-hou

In [69]:
# 위치 변경을 위해 'native-country_ Holand-Netherlands' 컬럼을 tmp 변수에 넣어줌

tmp = X_train['native-country_ Holand-Netherlands']

In [70]:
X_train.drop('native-country_ Holand-Netherlands', axis = 1, inplace=True)

In [71]:
X_train = pd.concat([X_train, tmp],axis=1)
X_train

age  fnlwgt  education-num  hours-per-week  workclass_ Federal-gov  \
no                                                                          
1       25  219199              7              40                       0   
2       39   52978             10              55                       0   
3       35  196899             13              50                       0   
4       64  135527             11              40                       0   
5       24   60783             10              70                       0   
...    ...     ...            ...             ...                     ...   
29301   20  100605              9              40                       0   
29302   21  372636              9              40                       0   
29303   18  258474              6              40                       0   
29304   33  157446              7              65                       0   
29305   65   94809              9              40                       0   

       workclass_ Local-gov  workclass_ Never-worked  workclass_ Private  \
no                                                                         
1                         0                        0                   1   
2                         0                        0                   1   
3                         0                        0                   1   
4                         0                        0                   1   
5                         0                        0                   1   
...                     ...                      ...                 ...   
29301                     0                        0                   1   
29302                     0                        0                   1   
29303                     0                        0                   0   
29304                     0                        0                   1   
29305                     0                        0                   0   

       workclass_ Self-emp-inc  workclass_ Self-emp-not-inc  ...  \
no                                                           ...   
1                            0                            0  ...   
2                            0                            0  ...   
3                            0                            0  ...   
4                            0                            0  ...   
5                            0                            0  ...   
...                        ...                          ...  ...   
29301                        0                            0  ...   
29302                        0                            0  ...   
29303                        0                            1  ...   
29304                        0                            0  ...   
29305                        0                            0  ...   

       native-country_ Vietnam  native-country_ Yugoslavia  \
no                                                           
1                            0                           0   
2                            0                           0   
3                            0                           0   
4                            0                           0   
5                            0                           0   
...                        ...                         ...   
29301                        0                           0   
29302                        0                           0   
29303                        0                           0   
29304                        0                           0   
29305                        0                           0   

       native-country_Others  capital-gain-group_Zero  \
no                                                      
1                          0                        1   
2                          0                        1   
3                          0                        1   
4                          0           

In [72]:
X_test

age  fnlwgt  education-num  hours-per-week  workclass_ Federal-gov  \
no                                                                          
29306   18  245274             10              16                       0   
29307   29   83003              9              40                       0   
29308   45   35136             13              40                       0   
29309   42   64631             13              40                       0   
29310   41  195821             16              40                       0   
...    ...     ...            ...             ...                     ...   
48838   45  116789              9              60                       0   
48839   48  185079              9              50                       0   
48840   63  117473             10              40                       0   
48841   18  150817              7              20                       0   
48842   31  341632             12              46                       0   

       workclass_ Local-gov  workclass_ Never-worked  workclass_ Private  \
no                                                                         
29306                     0                        0                   0   
29307                     0                        0                   1   
29308                     0                        0                   1   
29309                     0                        0                   0   
29310                     0                        0                   1   
...                     ...                      ...                 ...   
48838                     0                        0                   0   
48839                     0                        0                   1   
48840                     0                        0                   1   
48841                     0                        0                   1   
48842                     0                        0                   1   

       workclass_ Self-emp-inc  workclass_ Self-emp-not-inc  ...  \
no                                                           ...   
29306                        0                            0  ...   
29307                        0                            0  ...   
29308                        0                            0  ...   
29309                        0                            1  ...   
29310                        0                            0  ...   
...                        ...                          ...  ...   
48838                        0                            1  ...   
48839                        0                            0  ...   
48840                        0                            0  ...   
48841                        0                            0  ...   
48842                        0                            0  ...   

       native-country_ Vietnam  native-country_ Yugoslavia  \
no                                                           
29306                        0                           0   
29307                        0                           0   
29308                        0                           0   
29309                        0                           0   
29310                        0                           0   
...                        ...                         ...   
48838                        0                           0   
48839                        0                           0   
48840                        0                           0   
48841                        0                           0   
48842                        0                           0   

       native-country_Others  capital-gain-group_Zero  \
no                                                      
29306                      0                        0   
29307                      0                        0   
29308                      0                        0   
29309                      0           

### Scaler 만들기

In [73]:
from sklearn.preprocessing import StandardScaler

In [74]:
scaler = StandardScaler()
scaler.fit(X_train)

StandardScaler(copy=True, with_mean=True, with_std=True)

In [75]:
transform_X_train = scaler.transform(X_train)
transform_X_train

array([[-9.93857399e-01,  2.78921769e-01, -1.19948070e+00, ...,
        -1.56432881e-01, -1.52846987e-01, -5.84166364e-03],
       [ 2.47991359e-02, -1.29464937e+00, -3.23139826e-02, ...,
         6.39251795e+00, -1.52846987e-01, -5.84166364e-03],
       [-2.66245588e-01,  6.78134468e-02,  1.13485273e+00, ...,
        -1.56432881e-01, -1.52846987e-01, -5.84166364e-03],
       ...,
       [-1.50318567e+00,  6.50728019e-01, -1.58853627e+00, ...,
        -1.56432881e-01, -1.52846987e-01, -5.84166364e-03],
       [-4.11767950e-01, -3.05677882e-01, -1.19948070e+00, ...,
        -1.56432881e-01, -1.52846987e-01, -5.84166364e-03],
       [ 1.91658984e+00, -8.98646133e-01, -4.21369553e-01, ...,
        -1.56432881e-01, -1.52846987e-01, -5.84166364e-03]])

In [76]:
transform_X_test = scaler.transform(X_test)
transform_X_test

array([[-1.50318567,  0.52576704, -0.03231398, ..., -0.15643288,
        -0.15284699, -0.00584166],
       [-0.70281267, -1.01041048, -0.42136955, ..., -0.15643288,
        -0.15284699, -0.00584166],
       [ 0.46136622, -1.46355496,  1.13485273, ..., -0.15643288,
        -0.15284699, -0.00584166],
       ...,
       [ 1.77106748, -0.68409192, -0.03231398, ..., -0.15643288,
        -0.15284699, -0.00584166],
       [-1.50318567, -0.36843291, -1.1994807 , ..., -0.15643288,
        -0.15284699, -0.00584166],
       [-0.55729031,  1.43796326,  0.74579716, ..., -0.15643288,
        -0.15284699, -0.00584166]])

### 모델 생성하기

In [77]:
!pip install lightgbm

In [78]:
!pip install xgboost

In [79]:
from sklearn.model_selection import cross_val_score, train_test_split, KFold
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from lightgbm import LGBMClassifier
from xgboost import XGBClassifier
from sklearn.ensemble import RandomForestClassifier

In [80]:
lgb_model = LGBMClassifier(n_estimators = 1000, learning_rate=0.01, random_state=123)
lgb_model.fit(transform_X_train, y_train)

LGBMClassifier(boosting_type='gbdt', class_weight=None, colsample_bytree=1.0,
               importance_type='split', learning_rate=0.01, max_depth=-1,
               min_child_samples=20, min_child_weight=0.001, min_split_gain=0.0,
               n_estimators=1000, n_jobs=-1, num_leaves=31, objective=None,
               random_state=123, reg_alpha=0.0, reg_lambda=0.0, silent=True,
               subsample=1.0, subsample_for_bin=200000, subsample_freq=0)

In [81]:
xgb_model = XGBClassifier(n_estimators = 100, learning_rate=0.01, random_state = 123)
xgb_model.fit(transform_X_train, y_train)

C:\Users\rnjsx\anaconda3\lib\site-packages\xgboost\sklearn.py:892: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


[19:00:58] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.3.0/src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.


XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
              importance_type='gain', interaction_constraints='',
              learning_rate=0.01, max_delta_step=0, max_depth=6,
              min_child_weight=1, missing=nan, monotone_constraints='()',
              n_estimators=100, n_jobs=12, num_parallel_tree=1,
              objective='binary:logistic', random_state=123, reg_alpha=0,
              reg_lambda=1, scale_pos_weight=1, subsample=1,
              tree_method='exact', use_label_encoder=True,
              validate_parameters=1, verbosity=None)

In [82]:
rf_model = RandomForestClassifier(n_estimators=100, random_state= 11)
rf_model.fit(transform_X_train, y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=11, verbose=0,
                       warm_start=False)

### 교차 검증하기

Standard Scaler를 적용한 LightGBM 모델 kfold 교차 검증

정확도와 f1 구하기

In [83]:
kfold = KFold(n_splits=10, random_state=123, shuffle=True)

In [84]:
result1 = cross_val_score(lgb_model,
                         transform_X_train,
                         y_train,
                         cv = kfold
                        )

print(result1.mean())

0.8587957144004946


In [85]:
f1_result1 = cross_val_score(lgb_model, # lgb 모델
                         transform_X_train,
                         y_train,
                         cv = kfold,
                         scoring='f1'
                        )

print(f1_result1.mean())

0.6803504509092442


Standard Scaler를 적용한 XGBoost 모델 kfold 교차 검증

정확도와 f1 구하기

In [86]:
result2 = cross_val_score(xgb_model,
                          transform_X_train,
                          y_train,
                          cv = kfold
                         )

print(result2.mean())

[19:01:58] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.3.0/src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.


C:\Users\rnjsx\anaconda3\lib\site-packages\xgboost\sklearn.py:892: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


[19:02:01] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.3.0/src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[19:02:03] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.3.0/src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[19:02:05] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.3.0/src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[19:02:07] WARNING: C:/Users/Administrator/workspace/xgboost-win64_rel

In [87]:
f1_result2 = cross_val_score(xgb_model,
                          transform_X_train,
                          y_train,
                          cv = kfold,
                          scoring='f1'   
                         )

print(f1_result2.mean())

[19:02:19] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.3.0/src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[19:02:21] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.3.0/src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[19:02:23] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.3.0/src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[19:02:25] WARNING: C:/Users/Administrator/workspace/xgboost-win64_rel

Standard Scaler를 적용한 RandomForest 모델 kfold 교차 검증

정확도와 f1 구하기

In [88]:
result3 = cross_val_score(rf_model,
                          transform_X_train,
                          y_train,
                          cv = kfold
                         )

print(result3.mean())

0.8458289113780781


In [89]:
f1_result3 = cross_val_score(rf_model,
                          transform_X_train,
                          y_train,
                          cv = kfold,
                          scoring='f1'   
                         )

print(f1_result3.mean())

0.653523472475875


# Kaggle 제출용 파일 만들기

In [90]:
result = pd.read_csv('sample_submission.csv')
result

no  income
0      29306       0
1      29307       0
2      29308       0
3      29309       0
4      29310       0
...      ...     ...
19532  48838       1
19533  48839       0
19534  48840       1
19535  48841       0
19536  48842       0

[19537 rows x 2 columns]

In [91]:
pre = lgb_model.predict(transform_X_test)  # 모델로 예측

In [92]:
result['income'] = pre

In [93]:
result.to_csv('submission_capital2_med.csv', index=False)